tests/returntypes/gradients/test_pulse_gradient_new.py

```diff
@@ -0,0 +1,879 @@
+            qml.evolve(ham_single_q_const)(params, T)
+            return qml.expval(qml.PauliZ(0))
+
+        params = [jnp.array(0.4)]
+        grad = jax.jacobian(circuit)(params)
+        p = params[0] * T
+        exp_grad = -2 * jnp.sin(2 * p) * T
+        assert qml.math.allclose(grad, exp_grad, atol=tol, rtol=0.0)
+
+    @pytest.mark.parametrize("shots, tol", [(None, 1e-4), (100, 0.1), ([100, 99], 0.1)])
+    @pytest.mark.parametrize("num_samples", [1, 3])
+    def test_simple_qnode_probs(self, num_samples, shots, tol):
+        """Test that a simple qnode that returns an probabilities
+        can be differentiated with stoch_pulse_grad."""
+        import jax
+
+        jnp = jax.numpy
+        jax.config.update("jax_enable_x64", True)
+        dev = qml.device("default.qubit.jax", wires=1, shots=shots, prng_key=jax.random.PRNGKey(74))
+        T = 0.2
+        ham_single_q_const = qml.pulse.constant * qml.PauliY(0)
+
+        @qml.qnode(dev, interface="jax", diff_method=stoch_pulse_grad, num_samples=num_samples)
+        def circuit(params):
+            qml.evolve(ham_single_q_const)(params, T)
+            return qml.probs(wires=0)
+
+        params = [jnp.array(0.4)]
+        jac = jax.jacobian(circuit)(params)
+        p = params[0] * T
+        exp_jac = jnp.array([-1, 1]) * jnp.sin(2 * p) * T
+        assert qml.math.allclose(jac, exp_jac, atol=tol, rtol=0.0)
+
+    @pytest.mark.parametrize("shots, tol", [(None, 1e-4), (100, 0.1), ([100, 100], 0.1)])
+    @pytest.mark.parametrize("num_samples", [1, 3])
+    def test_simple_qnode_probs_expval(self, num_samples, shots, tol):
+        """Test that a simple qnode that returns an probabilities
+        can be differentiated with stoch_pulse_grad."""
+        import jax
+
+        jnp = jax.numpy
+        jax.config.update("jax_enable_x64", True)
+        dev = qml.device("default.qubit.jax", wires=1, shots=shots, prng_key=jax.random.PRNGKey(74))
+        T = 0.2
+        ham_single_q_const = qml.pulse.constant * qml.PauliY(0)
+
+        @qml.qnode(dev, interface="jax", diff_method=stoch_pulse_grad, num_samples=num_samples)
+        def circuit(params):
+            qml.evolve(ham_single_q_const)(params, T)
+            return qml.probs(wires=0), qml.expval(qml.PauliZ(0))
+
+        params = [jnp.array(0.4)]
+        jac = jax.jacobian(circuit)(params)
+        p = params[0] * T
+        exp_jac = (jnp.array([-1, 1]) * jnp.sin(2 * p) * T, -2 * jnp.sin(2 * p) * T)
+        if hasattr(shots, "len"):
+            for j_shots, e_shots in zip(jac, exp_jac):
+                for j, e in zip(j_shots, e_shots):
+                    assert qml.math.allclose(j[0], e, atol=tol, rtol=0.0)
+        else:
+            for j, e in zip(jac, exp_jac):
+                assert qml.math.allclose(j[0], e, atol=tol, rtol=0.0)
+
+    # TODO: ParametrizedEvolution + grad transform is currently not JIT-compatible, see #3881
+    @pytest.mark.xfail
+    @pytest.mark.parametrize("num_samples", [1, 3])
```

Comment:
I will try to give it a look!

Reply:
```
self.t = jnp.array([0, t] if qml.math.ndim(t) == 0 else t, dtype=float)
```
This creates leaked tracers

I made it work by applying two changes:

1. jit_grad = jax.jit(jax.grad(circuit), static_argnums=[1])(params, T)
2. self.t = numpy.array([0, t])

I think you probably change the parametrized evolution if you want to jit the time parameter as well.

Reply:
In principle one should be able to jit the time argument in the 'odeint' solver. It doesnt work with just numpy t (and no static argnums)?

Reply:
Indeed :+1: